modules/imgproc/src/pyramids.cpp

```diff
@@ -753,109 +892,14 @@ pyrDown_( const Mat& _src, Mat& _dst, int borderType )
-                    {
-                        const T* s = src + x*2;
-                        WT t0 = s[0] * 6 + (s[-2] + s[2]) * 4 + s[-4] + s[4];
-                        WT t1 = s[1] * 6 + (s[-1] + s[3]) * 4 + s[-3] + s[5];
-                        row[x] = t0; row[x + 1] = t1;
-                    }
-                }
-                else if( cn == 3 )
-                {
-                    x += PyrDownVecH<T, WT, 3>(src + x * 2 - 6, row + x, width0 - x);
-                    for( ; x < width0; x += 3 )
-                    {
-                        const T* s = src + x*2;
-                        WT t0 = s[0]*6 + (s[-3] + s[3])*4 + s[-6] + s[6];
-                        WT t1 = s[1]*6 + (s[-2] + s[4])*4 + s[-5] + s[7];
-                        WT t2 = s[2]*6 + (s[-1] + s[5])*4 + s[-4] + s[8];
-                        row[x] = t0; row[x+1] = t1; row[x+2] = t2;
-                    }
-                }
-                else if( cn == 4 )
-                {
-                    x += PyrDownVecH<T, WT, 4>(src + x * 2 - 8, row + x, width0 - x);
-                    for( ; x < width0; x += 4 )
-                    {
-                        const T* s = src + x*2;
-                        WT t0 = s[0]*6 + (s[-4] + s[4])*4 + s[-8] + s[8];
-                        WT t1 = s[1]*6 + (s[-3] + s[5])*4 + s[-7] + s[9];
-                        row[x] = t0; row[x+1] = t1;
-                        t0 = s[2]*6 + (s[-2] + s[6])*4 + s[-6] + s[10];
-                        t1 = s[3]*6 + (s[-1] + s[7])*4 + s[-5] + s[11];
-                        row[x+2] = t0; row[x+3] = t1;
-                    }
-                }
-                else
-                {
-                    for( ; x < width0; x++ )
-                    {
-                        int sx = tabM[x];
-                        row[x] = src[sx]*6 + (src[sx - cn] + src[sx + cn])*4 +
-                            src[sx - cn*2] + src[sx + cn*2];
-                    }
-                }
-
-                // tabR
-                for (int x_ = 0; x < dsize.width; x++, x_++)
-                {
-                    row[x] = src[tabR[x_+cn*2]]*6 + (src[tabR[x_+cn]] + src[tabR[x_+cn*3]])*4 +
-                        src[tabR[x_]] + src[tabR[x_+cn*4]];
-                }
-            } while (0);
-        }
-
-        // do vertical convolution and decimation and write the result to the destination image
-        for (int k = 0; k < PD_SZ; k++)
-            rows[k] = buf + ((y*2 - PD_SZ/2 + k - sy0) % PD_SZ)*bufstep;
-        row0 = rows[0]; row1 = rows[1]; row2 = rows[2]; row3 = rows[3]; row4 = rows[4];
+    int *tabLPtr = tabL;
+    int *tabRPtr = tabR;
 
-        int x = PyrDownVecV<WT, T>(rows, dst, dsize.width);
-        for (; x < dsize.width; x++ )
-            dst[x] = castOp(row2[x]*6 + (row1[x] + row3[x])*4 + row0[x] + row4[x]);
-    }
+    cv::parallel_for_(Range(0,dsize.height), cv::PyrDownInvoker<CastOp>(_src, _dst, borderType, &tabRPtr, &tabM, &tabLPtr));
```

Comment:
Consider passing granularity (nstripes) parameter - there is overhead from ring buffer initialization

Reply:
Just pass `parallel_for_()` `nstripes` parameter equal to `cv::getNumThreads()`
(there is no direct support for granularity in OpenCV)

Reply:
Done!